designs/extensible-cli-and-scaffolding-plugins.md

```diff
@@ -0,0 +1,144 @@
+# Extensible CLI and Scaffolding Plugins
+
+## Overview 
+I would like for Kubebuilder to become more extensible, such that it could be imported and used as a library in other projects. Specifically, I'm looking for a way to use Kubebuilder's existing CLI and scaffolding for Go projects, but to also be able to augment Kubebuilder's project versions with other custom project versions so that I can support the Kubebuilder workflow with non-Go operators (e.g. operator-sdk's Ansible and Helm-based operators).
+
+The idea is for Kubebuilder to define one or more plugin interfaces that can be used to drive what the `init`, `create api` and `create webhooks` subcommands do and to add a new `cli` package that other projects can use to integrate out-of-tree plugins with the Kubebuilder CLI in their own projects.
+
+## Related issues and PRs
+
+* [#1148](https://github.com/kubernetes-sigs/kubebuilder/pull/1148)
+* [#1171](https://github.com/kubernetes-sigs/kubebuilder/pull/1171)
+* Possibly [#1218](https://github.com/kubernetes-sigs/kubebuilder/issues/1218)
+
+## Prototype implementation
+https://github.com/joelanford/kubebuilder-exp
+
+## Plugin interfaces
+
+### Required
+Each plugin would minimally be required to implement the `Plugin` interface.
+
+```go
+type Plugin interface {
+    // Version is the project version that this plugin implements.
+    // For example, Kubebuilder's Go v2 plugin implementation would return "2"
+    Version() string
```

Comment:
WDYT about we have here the type as well? (ansible, helm, go)?

Reply:
See https://github.com/kubernetes-sigs/kubebuilder/pull/1250#issuecomment-564259130 and below for my thoughts.

Reply:
I was thinking that kubebuilder or sdk are equals when is the go type, so I mean it doesn't really matter since the scaffold will be the same and what may change are the top customizations. But may I missing some point.

Reply:
Things we seem to need from the base `Plugin` interface:
- Method to get a version string (`Version()`)
  - Always an integer as string.
- Method to get an identifier for a plugin (`Name()`)
  - Should this identifier format be up the implementer or be `name:vX`?

If we go with a self-specified identifier, do we care about prevent naming conflicts between plugins?

Reply:
+1 on name and version

We should not force folks to determine if they are are type or a builder, a name for the given plugin that is being used and a version of that plugin will give us the flexibility we need  IMO. 

if `Version` is always an int, then can we make it an unsigned int type to get compile-time guarantees?